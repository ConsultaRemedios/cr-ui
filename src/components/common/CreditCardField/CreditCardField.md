```js
new Vue({
  data(){
    return { creditCardNumber: '' };
  },
  methods: {
    onChange(e) {
      this.creditCardNumber = e.value
    }
  },
  template: `
    <CreditCardField @change="onChange" :value="creditCardNumber"/>
  `,
});
```
