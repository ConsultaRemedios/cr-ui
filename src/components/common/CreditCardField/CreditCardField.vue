<template>
  <div :class="$style.wrapper">
    <BaseInput
      :name="name"
      type="tel"
      @change="onChange"
      @blur="onBlur"
      :value="currentValue"
      :has-error="hasError"
      :custom-style="{ paddingRight: '46px' }"
      v-bind="$attrs"
    />

    <img
      v-if="brandIcon"
      :class="$style.brandIcon"
      :src="brandIcon"
    />
  </div>
</template>

<script>
  import getCreditCardInfo from './../../../helpers/credit-card-info';
  import BaseInput from '../BaseInput';

  import visa from './../../../assets/credit-card-brands/visa.png';
  import mastercard from './../../../assets/credit-card-brands/mastercard.png';
  import discover from './../../../assets/credit-card-brands/discover.png';
  import jcb from './../../../assets/credit-card-brands/jcb.png';
  import americanExpress from './../../../assets/credit-card-brands/american-express.png';
  import dinersClub from './../../../assets/credit-card-brands/diners-club.png';
  import hipercard from './../../../assets/credit-card-brands/hipercard.png';
  import elo from './../../../assets/credit-card-brands/elo.png';
  import maestro from './../../../assets/credit-card-brands/maestro.png';

  export default {
    inheritAttrs: false,
    name: 'CreditCardField',

    props: {
      name: {
        type: String,
        default: '',
      },

      value: {
        type: String,
        default: '',
      },

      hasError: {
        type: Boolean,
        default: false,
      },
    },

    data() {
      return { brand: '', currentValue: this.value };
    },

    components: { BaseInput },

    watch: {
      value(newValue) {
        this.currentValue = newValue;
      },
    },

    methods: {
      onBlur(e) {
        const { formattedValue, brand } = getCreditCardInfo(e.value);

        this.$emit('blur', {
          ...e,
          value: formattedValue,
          brand,
        });
      },

      onChange(e) {
        const { formattedValue, brand } = getCreditCardInfo(e.value);

        this.currentValue = e.value;
        this.brand = brand;

        this.$nextTick(() => {
          this.currentValue = formattedValue;

          if (this.value === formattedValue) return;

          this.$emit('change', {
            ...e,
            value: formattedValue,
            brand,
          });
        });
      },
    },

    computed: {
      brandIcon() {
        const icons = {
          'american-express': americanExpress,
          'diners-club': dinersClub,
          discover,
          elo,
          hipercard,
          jcb,
          maestro,
          mastercard,
          visa,
        };

        return icons[this.brand];
      },
    },
  };
</script>

<style module>
  .wrapper {
    position: relative;
  }

  .brandIcon {
    position: absolute;
    background: #CCC;
    text-align: center;
    text-transform: uppercase;
    top: 50%;
    right: 12px;
    transform: translateY(-50%);
  }
</style>
