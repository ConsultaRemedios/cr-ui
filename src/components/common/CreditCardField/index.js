export { default } from './CreditCardField.vue';
