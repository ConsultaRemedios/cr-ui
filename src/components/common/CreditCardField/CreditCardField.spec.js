import { mount, shallowMount } from '@vue/test-utils';
import { BaseInput } from 'cr-ui';
import snapshotDiff from 'snapshot-diff';
import Vue from 'vue';
import { fireEvent, render } from 'vue-testing-library';
import getCreditCardInfo from './../../../helpers/credit-card-info';
import CreditCardField from './CreditCardField.vue';

jest.mock('./../../../helpers/credit-card-info', () => jest.fn());

const createEvent = (value = 'abc54799886282def08503ghj') => ({ target: { value } });

describe('CreditCardField', () => {
  let html;
  let props;

  afterEach(() => {
    document.body.innerHTML = '';

    props = {
      name: 'card',
    };
  });

  describe('when it is empty', () => {
    it('matches the snapshot', () => {
      const wrapper = shallowMount(CreditCardField, { propsData: props });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('when a card number is typed', () => {
    it('matches the snapshot', () => {
      getCreditCardInfo.mockImplementation(() => ({
        brand: 'american-express',
        formattedValue: '5479 9886 2820 8503'
      }));

      const wrapper = shallowMount(CreditCardField, { propsData: props });
      const wrapper2 = shallowMount(CreditCardField, { propsData: props });

      wrapper2.find(BaseInput).vm.$emit('change', { value: '54799886282def08503' });

      expect.assertions(1);

      return Vue.nextTick().then(() => {
        expect(snapshotDiff(wrapper.element, wrapper2.element)).toMatchSnapshot();
      })
    });
  });

  it('formats input properly', () => {
    getCreditCardInfo.mockImplementation(() => ({
      brand: 'american-express',
      formattedValue: '5479 9886 2820 8503'
    }));

    const { getByDisplayValue } = render(CreditCardField, { props });
    fireEvent.input(document.querySelector('[name="card"]'), createEvent());

    expect.assertions(1);

    return Vue.nextTick().then(() => {
      expect(document.querySelector('[name="card"]').value).toEqual('5479 9886 2820 8503');
    })
  });

  describe('brand images', () => {
    it('renders "american express" brand image', () => {
      getCreditCardInfo.mockImplementation(() => ({
        brand: 'american-express'
      }));

      render(CreditCardField, { props });
      fireEvent.input(document.querySelector('[name="card"]'), createEvent());

      return Vue.nextTick().then(() => {
        expect(document.querySelector('img').src).toEqual('http://localhost/american-express.png');
      });
    });

    it('renders "diners-club" brand image', () => {
      getCreditCardInfo.mockImplementation(() => ({
        brand: 'diners-club'
      }));

      render(CreditCardField, { props });
      fireEvent.input(document.querySelector('[name="card"]'), createEvent());

      return Vue.nextTick().then(() => {
        expect(document.querySelector('img').src).toEqual('http://localhost/diners-club.png');
      });
    });

    it('renders "discover" brand image', () => {
      getCreditCardInfo.mockImplementation(() => ({
        brand: 'discover'
      }));

      render(CreditCardField, { props });
      fireEvent.input(document.querySelector('[name="card"]'), createEvent());

      return Vue.nextTick().then(() => {
        expect(document.querySelector('img').src).toEqual('http://localhost/discover.png');
      });
    });

    it('renders "elo" brand image', () => {
      getCreditCardInfo.mockImplementation(() => ({
        brand: 'elo'
      }));

      render(CreditCardField, { props });
      fireEvent.input(document.querySelector('[name="card"]'), createEvent());

      return Vue.nextTick().then(() => {
        expect(document.querySelector('img').src).toEqual('http://localhost/elo.png');
      });
    });

    it('renders "hipercard" brand image', () => {
      getCreditCardInfo.mockImplementation(() => ({
        brand: 'hipercard'
      }));

      render(CreditCardField, { props });
      fireEvent.input(document.querySelector('[name="card"]'), createEvent());

      return Vue.nextTick().then(() => {
        expect(document.querySelector('img').src).toEqual('http://localhost/hipercard.png');
      });
    });

    it('renders "jcb" brand image', () => {
      getCreditCardInfo.mockImplementation(() => ({
        brand: 'jcb'
      }));

      render(CreditCardField, { props });
      fireEvent.input(document.querySelector('[name="card"]'), createEvent());

      return Vue.nextTick().then(() => {
        expect(document.querySelector('img').src).toEqual('http://localhost/jcb.png');
      });

    });

    it('renders "maestro" brand image', () => {
      getCreditCardInfo.mockImplementation(() => ({
        brand: 'maestro'
      }));

      render(CreditCardField, { props });
      fireEvent.input(document.querySelector('[name="card"]'), createEvent());

      return Vue.nextTick().then(() => {
        expect(document.querySelector('img').src).toEqual('http://localhost/maestro.png');
      });
    });

    it('renders "mastercard" brand image', () => {
      getCreditCardInfo.mockImplementation(() => ({
        brand: 'mastercard'
      }));

      render(CreditCardField, { props });
      fireEvent.input(document.querySelector('[name="card"]'), createEvent());

      return Vue.nextTick().then(() => {
        expect(document.querySelector('img').src).toEqual('http://localhost/mastercard.png');
      });
    });

    it('renders "visa" brand image', () => {
      getCreditCardInfo.mockImplementation(() => ({
        brand: 'visa'
      }));

      render(CreditCardField, { props });
      fireEvent.input(document.querySelector('[name="card"]'), createEvent());

      return Vue.nextTick().then(() => {
        expect(document.querySelector('img').src).toEqual('http://localhost/visa.png');
      });
    });
  });

  it('does not emit change event when the currentValue is equal to the value prop', () => {
    getCreditCardInfo.mockImplementation(() => ({
      brand: 'american-express',
      formattedValue: '5479 9886 2820 8503'
    }));

    const propsData = { ...props, value: '5479 9886 2820 8503' };

    const wrapper = mount(CreditCardField, { propsData });

    wrapper.find('input').element.value = 'abc54799886282def08503ghj';
    wrapper.find('input').trigger('input');

    wrapper.find('input').element.value = 'a54799886282def08503g';
    wrapper.find('input').trigger('input');

    wrapper.find('input').element.value = '54799886282def08503';
    wrapper.find('input').trigger('input');

    return Vue.nextTick().then(() => {
      expect(wrapper.emitted().change).toBeFalsy();
    });
  });

  it('updates input value when the value prop is changed', () => {
    const wrapper = mount(CreditCardField, { propsData: props });

    wrapper.setProps({ value: '5478' });

    expect(wrapper.find('input').element.value).toEqual('5478');
  });

  it('emits "blur" event with the proper params', () => {
    const wrapper = mount(CreditCardField, { propsData: props });
    jest.spyOn(wrapper.vm, '$emit');
    wrapper.find('input').trigger('blur');

    const emitted = wrapper.vm.$emit.mock.calls[0];

    expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);
    expect(emitted[0]).toEqual('blur');
    expect(emitted[1]).toMatchObject({ name: 'card', value: '5479 9886 2820 8503' });
    expect(emitted[1].event.target).toEqual(wrapper.find('input').element);
  });
});
