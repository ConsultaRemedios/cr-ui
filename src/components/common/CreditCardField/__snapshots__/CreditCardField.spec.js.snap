// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditCardField when a card number is typed matches the snapshot 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -6,10 +6,13 @@
      customstyle="[object Object]"
      model=""
      name="card"
      size="medium"
      type="tel"
-     value=""
+     value="5479 9886 2820 8503"
    />
     
-   <!---->
+   <img
+     class="brandIcon"
+     src="american-express.png"
+   />
  </div>
`;

exports[`CreditCardField when it is empty matches the snapshot 1`] = `
<div
  class="wrapper"
>
  <baseinput-stub
    classes=""
    customstyle="[object Object]"
    model=""
    name=""
    size="medium"
    type="tel"
    value=""
  />
   
  <!---->
</div>
`;
